### PR TITLE
Make umount fuse properly when killing the process

### PR DESCRIPTION
**Pull request by yabola**
It’s just a draft for fast review in #13239

### CODE REVIEW

**Comment by apc999** [comment]
@yabola  are you still working on this PR?

**Comment by yabola** [comment]
@apc999 right, but I want to confirm whether the idea of intercepting kill requests is appropriate or any other new idea on this issue.I will continue to complete this PR if it is an appropriate way.

**Comment by LuQQiu** [comment]
@apc999 I remember that you took a look before and think that another function other than umount should be used?

**Comment by LuQQiu** [comment]
@yabola 
We don't have a pretty good idea of how to properly umount fuse, so would like to know what's the behavior before and after your changes? Have you tested your changes out in standalone fuse and worker fuse? what's the behavior looks like.

**Comment by yabola** [comment]
@LuQQiu 
My test step:
set properties` alluxio.fuse.umount.timeout=5s` ->  mount fuse dir ->
vi fuse_dir/b.txt ->  find fuse JVM Pid -> kill pid

Before this PR:
fuse JVM is killed.  the file which i vi before show`"fuse_dir/b.txt" E212: Can't open file for writing` (i use :wq command)

After this PR:
fuse JVM is not killed.The fuse JVM acts normally as before.

now this PR only catch kill signal 15 (kill pid). it hadn't catch other kill signal like kill -9.

**Comment by LuQQiu** [comment]
@yabola this behavior looks good for the standalone fuse. When fuse is in the worker process, what's the behavior before and after your changes?
Basically we would want a solution that properly umount fuse in both standalone way and worker fuse. Ideally, when fuse device is busy, standalone fuse umount should be tricked as noop. Fuse in worker is a little tricky, we don't want the worker process components to be killed while the fuse mount is still there leaving the fuse mount in unusable state. We would rather force the fuse to be unmounted or prevent the whole worker process from exiting (trick worker process kill as noop). Not sure if it's possible

**Comment by yabola** [comment]
@LuQQiu 
When fuse is in the worker process, it can act like standalone fuse . when use kill AlluxioWorker pid command, worker process kill as noop.  
But    `alluxio-stop.sh worker` command will kill -9 pid if kill pid failed after 120s. (https://github.com/Alluxio/alluxio/blob/master/bin/alluxio#L101). this command will finally kill worker process.
so if we need change `kill -9 worker pid` behavior here?

**Comment by yabola** [comment]
@LuQQiu @apc999 
what's your opinions about that?

**Comment by LuQQiu** [comment]
@yabola 
```
When fuse is in the worker process, it can act like standalone fuse . when use kill AlluxioWorker pid command, worker process kill as noop.
```
Can you elaborate more on this line?  why worker process kill will be treated as noop? Fuse will directly capture the kill signal and prevent that? it will prevent Alluxio worker stop all other functionalities?

**Comment by yabola** [comment]
@LuQQiu 
 worker and  fuse is in one JVM when embedded-fuse worker started.`SignalHandler` is also registered in this JVM. So embedded-fuse worker JVM will listen to kill signal as standalone fuse do. when user use kill pid command, it will act as noop if fuse device is busy. However, it can't catch kill -9 pid signal.

**Comment by LuQQiu** [comment]
As long as we get the desired behavior, your design and initial draft LGTM.
Thanks for the detailed description.
One small question is about the umount(force), the force parameter doesn't look like doing anything useful. 
For alluxio-stop.sh worker, we can add a force parameter, by default turn on to sync with the previous behavior, but user can pass in force=false to not trigger kill -9

**Comment by yabola** [comment]
@LuQQiu fine,I will complete my PR later~

**Comment by yabola** [comment]
Sorry for late. I have updated alluxio shell command as you said. 
This is my test evidence.


1. 
mount fuse dir in worker
```
alluxio.worker.fuse.enabled=true
alluxio.worker.fuse.mount.alluxio.path=/fuse_data
alluxio.worker.fuse.mount.point=/Users/chenliang.lu/soft/alluxio_data/fuse_data
```
2.  
open operation on mount file
`less /Users/chenliang.lu/soft/alluxio_data/fuse_data/b.txt`
image

3. 
 `sh bin/alluxio-stop.sh -s worker ` ---> won't kill worker
error message:
```
 Error, process [70125] did not complete after 120 seconds, failing to kill it.
Successfully Killed 0 process(es) successfully on ChengliangLudeMacBook-Pro.local
Failed to kill 1 process(es) due to soft kill option on ChengliangLudeMacBook-Pro.local
```
image

 4.
 `sh bin/alluxio-stop.sh  worker ` ---> will kill worker forcily finally
 
image
image

**Comment by yabola** [comment]
Test Evidence in `alluxio-fuse`shell command：
When using the umount command, the fuse application does not respond.The AlluxioFuse process still exists and can be used normally.

image

**Comment by LuQQiu** [comment]
Awesome!! @yabola Thanks for doing all the testing!!
The behavior LGTM!

File: tests/src/test/java/alluxio/client/fuse/JNIFuseIntegrationTest.java [hunk 1]
```diff
@@ -41,6 +41,6 @@ public void mountFuse(FileSystem fileSystem, String mountPoint, String alluxioRo
 
   @Override
   public void umountFuse(String mountPath) throws Exception {
-    mFuseFileSystem.umount();
+    mFuseFileSystem.umount(false);
```
Comment (LuQQiu) [inline tests/src/test/java/alluxio/client/fuse/JNIFuseIntegrationTest.java:44]: why false here? will it cause the test to leave the fuse process running while closing?

File: bin/alluxio [hunk 2]
```diff
@@ -79,31 +79,50 @@ EOF
 }
 
 function killAll {
-  if [[ $# -ne 1 ]]; then
-    echo "Usage: alluxio killAll <WORD>" >&2
+  if [[ $# -gt 2  ]]; then
+    echo "Usage: alluxio killAll [-s] <WORD>" >&2
```
Comment (LuQQiu) [inline bin/alluxio:83]: Did the `-s` only apply to AlluxioWorker?
In this usage, i will feel -s the soft kill option apply to all processes but looks like it only use in AlluxioWorker?

**Review by LuQQiu — COMMENTED** [review]
@yabola Can you help fix the `integration/fuse/bin/alluxio-fuse umount` output message while saying that the fuse is mounted but actually fuse is not mounted because fuse divice is busy?

Another question is about the `-s` soft kill option. I will prefer to apply it to all processes not only AlluxioWorker. @apc999 @ggezer what's your advice about adding a soft kill option that do not force all processes to be killed? do not kill -9 if kill -15 doesn't work

File: tests/src/test/java/alluxio/client/fuse/JNIFuseIntegrationTest.java (hunk 1 above)
Comment (yabola) [reply tests/src/test/java/alluxio/client/fuse/JNIFuseIntegrationTest.java:44]: done.

**Comment by yabola** [comment]
@LuQQiu 
My test evidence in alluxio-fuse shell command:
when device is busy,fail to kill:
image
when device is free, successfully kill it:
image

I am not sure if `-s` should apply to all processes. If it needs, I'm willing to update it.

File: bin/alluxio [hunk 3]
```diff
@@ -79,31 +79,50 @@ EOF
...
     local cnt=${maxWait}
+    fail_to_kill=false
     while kill -0 ${pid} > /dev/null 2>&1; do
       if [[ ${cnt} -gt 1 ]]; then
         # still not dead, wait
         cnt=$(expr ${cnt} - 1)
         sleep 1
+      elif [[ "${soft_kill_option}" = true ]] ; then
+        # waited long enough, give up killing the process
+        echo "Error, process ["${pid}"] did not complete after "${maxWait}" seconds, failing to kill it." >&2
```
Comment (LuQQiu) [inline bin/alluxio:107]:
```suggestion
        echo "Process ${pid} did not complete after "${maxWait}" seconds, failing to kill it." >&2
```

File: bin/alluxio [hunk 4]
```diff
@@ -79,31 +79,50 @@ EOF
...
+    if [[ "$fail_to_kill" = true ]]; then
+      failed_count=$(expr ${failed_count} + 1)
+    else
+      success_count=$(expr ${success_count} + 1)
+    fi
   done
-  echo "Killed ${count} process(es) on $(hostname)"
+  echo "Successfully Killed ${success_count} process(es) successfully on $(hostname)"
+  if [[ $failed_count -gt 0 ]]; then
+    echo "Failed to kill ${failed_count} process(es) due to soft kill option on $(hostname)"
```
Comment (LuQQiu) [inline bin/alluxio:124]:
```suggestion
    echo "Soft kill option -s specified, failed to kill ${failed_count} process(es) on $(hostname)"
```

File: integration/fuse/bin/alluxio-fuse [hunk 5]
```diff
@@ -88,9 +88,22 @@ umount_fuse() {
...
+    local maxWait=30
+    local cnt=${maxWait}
+    while kill -0 ${fuse_pid} > /dev/null 2>&1; do
+      if [[ ${cnt} -gt 1 ]]; then
+        # still not dead, wait
+        cnt=$(expr ${cnt} - 1)
+        sleep 1
+      else
+        # waited long enough, give up umount
+        echo "Error, fuse process ["${fuse_pid}"] did not complete after "${maxWait}" seconds, failing to umount fuse." >&2
```
Comment (LuQQiu) [inline integration/fuse/bin/alluxio-fuse:101]:
```suggestion
        echo "Failed to umount fuse process ["${fuse_pid}"] after waiting for "${maxWait}" seconds." >&2
```

File: integration/fuse/src/main/java/alluxio/fuse/FuseSignalHandler.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import org.slf4j.LoggerFactory;
+import sun.misc.Signal;
+import sun.misc.SignalHandler;
+
+/**
+ * Respond to the kill command when in mount fuse JVM.
+ * If it does not respond, the subsequent shutdown hook will not be triggered.
+ */
+public class FuseSignalHandler implements SignalHandler {
+
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/FuseSignalHandler.java:26]: Remove extra empthy line

File: integration/fuse/src/main/java/alluxio/fuse/FuseSignalHandler.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(FuseSignalHandler.class);
+
+  /**
+   * Use to umount Fuse application during stop.
+   */
+  private FuseUmountable mFuseUmountable;
+
+  /**
+   * init.
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/FuseSignalHandler.java:35]:
```suggestion
   * Constructs the new {@link FuseSingalHandler}.
   *
```

File: integration/fuse/src/main/java/alluxio/fuse/FuseUmountable.java [hunk 8]
```diff
@@ -11,12 +11,16 @@
...
+import alluxio.jnifuse.FuseException;
+
 /**
  * An interface for unmounting the mounted Fuse applications.
  */
 public interface FuseUmountable {
   /**
    * Unmounts the Fuse application.
+   * @param force when the umount times out due to the device is busy,
+   *              it will throw exception (catched by {@link FuseSignalHandler})
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/FuseUmountable.java:23]:
```
   * @param force true to throw an exception and treat umount as a no-op 
   *            when umount timeouts due to the fuse device is busy.
```
modify the previous line to fit the indentation codestyle

**Review by LuQQiu — COMMENTED** [review]
Left some minor comments. For the -s, can you add a description in bin/alluxio getUsage and telling that it is only used by AlluxioWorker.

**Comment by LuQQiu** [comment]
can you help fix the conflicts as well? thanks!

**Comment by LuQQiu** [comment]
@yabola Please modify the bin/alluxio-stop.sh usage as well, and add some related docs on the Alluxio POSIX API doc?

**Comment by ggezer** [comment]
> @yabola Can you help fix the `integration/fuse/bin/alluxio-fuse umount` output message while saying that the fuse is mounted but actually fuse is not mounted because fuse divice is busy?
> 
> Another question is about the `-s` soft kill option. I will prefer to apply it to all processes not only AlluxioWorker. @apc999 @ggezer what's your advice about adding a soft kill option that do not force all processes to be killed? do not kill -9 if kill -15 doesn't work

@LuQQiu I think that's fair. However, I'd like to get that done outside of this PR. ( Ideally this PR just rebases after that PR is merged.)

**Comment by LuQQiu** [comment]
Sounds good @ggezer, do you prefer @yabola to work on adding the -s to all processes or other engineers?

**Comment by ggezer** [comment]
> Sounds good @ggezer, do you prefer @yabola to work on adding the -s to all processes or other engineers?

Whoever has bandwidth really :D 
@yabola We'd appreciate if you could isolate this change into a separate PR complete with all processes.

**Comment by yabola** [comment]
> > Sounds good @ggezer, do you prefer @yabola to work on adding the -s to all processes or other engineers?
> 
> Whoever has bandwidth really :D
> @yabola We'd appreciate if you could isolate this change into a separate PR complete with all processes.

@ggezer @LuQQiu Do you mean first pull a PR for `-s` option, then  to compete umount fuse option PR?  I am willing to compete it~~

File: bin/alluxio [hunk 9]
```diff
@@ -79,31 +79,50 @@ EOF
...
     done
-    count=$(expr ${count} + 1)
+    if [[ "$fail_to_kill" = true ]]; then
+      failed_count=$(expr ${failed_count} + 1)
+    else
+      success_count=$(expr ${success_count} + 1)
+    fi
   done
-  echo "Killed ${count} process(es) on $(hostname)"
+  echo "Successfully Killed ${success_count} process(es) successfully on $(hostname)"
```
[the remark this answers is not in the document]
Comment (yabola) [reply bin/alluxio]: done

File: bin/alluxio (hunk 3 above)
Comment (yabola) [reply bin/alluxio]: done

File: bin/alluxio (hunk 4 above)
Comment (yabola) [reply bin/alluxio]: done

File: integration/fuse/bin/alluxio-fuse (hunk 5 above)
Comment (yabola) [reply integration/fuse/bin/alluxio-fuse]: done

File: integration/fuse/src/main/java/alluxio/fuse/FuseSignalHandler.java (hunk 6 above)
Comment (yabola) [reply integration/fuse/src/main/java/alluxio/fuse/FuseSignalHandler.java]: done

File: integration/fuse/src/main/java/alluxio/fuse/FuseUmountable.java (hunk 8 above)
Comment (yabola) [reply integration/fuse/src/main/java/alluxio/fuse/FuseUmountable.java]: done

File: integration/fuse/src/main/java/alluxio/fuse/FuseSignalHandler.java (hunk 7 above)
Comment (yabola) [reply integration/fuse/src/main/java/alluxio/fuse/FuseSignalHandler.java]: done

**Comment by yabola** [comment]
I have updated my PR to add soft kill option to all process. I will separate PR into two PR.

**Pull request closed without merging**